spec/loophp/collection/CollectionSpec.php

```diff
@@ -1139,6 +1139,25 @@ static function ($value) {
             ->shouldIterateAs([true]);
     }
 
+    public function it_can_flatMap(): void
+    {
+        $this::fromIterable([1, 2, 3])
+            ->flatMap(static fn (int $item): array => [$item + $item, $item * $item])
+            ->shouldIterateAs([2, 1, 4, 4, 6, 9]);
+
+        $this::fromIterable([1, 2, 3])
+            ->flatMap(static fn (int $item): array => [[$item + $item], [$item * $item]])
+            ->shouldIterateAs([[2], [1], [4], [4], [6], [9]]);
+
+        $this::fromIterable(['foo' => 'f', 'bar' => 'b'])
+            ->flatMap(static fn (string $item, string $key): array => [$item => $key])
+            ->shouldIterateAs([0 => 'foo', 1 => 'bar']);
```

Comment:
@drupol I'd like to clarify this case because I think this initial implementation is a bit faulty.

Here I would expect to get `['f' => 'foo', 'b' => 'bar']`, however, because of the usage of `Normalize`, the string keys get replaced with integer ones. I don't think `flatMap` should change the type of keys, it's not something that this operation typically does.

I reckon we would need to use `Pack` and `Unpack` in such a way that the keys are maintained, right? I tried a couple of things but could not get it working properly as the map callback needs to be changed. What do you think?

Reply:
Hi Alex!

Yes, we can safely remove the `normalize` I have no objection against it.

Reply:
the problem is just by removing it the operation doesn't work 😄 . Because the indexes are the same, so `Collection::fromIterable([1, 2, 3])->flatMap(static fn (int $item): array => [$item * $item])` will be `[0 => 9]` instead of `[1, 4, 9]`

Reply:
Let me get back to you after family lunch and I'll check with my computer.

Reply:
Ok here's my thoughts.

I think we should, at all cost try to keep the association `[$key, $value]`, even if this is absurd... but this is how PHP is meant to be used.
This library try to keep this association almost everywhere, despite the fact that this could be discussed.

Why? because the user is able to use methods `::pack()` or `::wrap()` before calling this operation.
Using such methods, will change the structure of the data passed in the subsequent calls, but it's the user who's responsible of that, so he needs to act accordingly.

Anyway, the solution here is very simple.

If you really want to deal with keys, then your example needs to be updated like this:

```php
Collection::fromIterable([1, 2, 3])
    ->flatMap(static fn (int $item, int $key): array => [$key => $item * $item]);
```

The `::map()` method is able to deal with keys and values, so you have to provide it.

Let me know what you think.

Reply:
Hmm at first I was a bit confused by this but I now see what you mean. It's best to not include `normalize` in this operation because it restricts the way it can be used when it comes to string keys. And when needed, the user can call that separately. These two would then be equivalent:

```PHP
Collection::fromIterable([1, 2, 3])
    ->flatMap(static fn (int $item, int $key): array => [$key => $item * $item]);

Collection::fromIterable([1, 2, 3])
    ->flatMap(static fn (int $item, int $key): array => [$item * $item])
    ->normalize();
```

Also this will work without normalize:

```PHP
$this::fromIterable(['foo' => 'f', 'bar' => 'b'])
            ->flatMap(static fn (string $item, string $key): array => [$item => $key])
            ->shouldIterateAs(['f' => 'foo', 'b' => 'bar']);

Reply:
Exactly, you got my point.